README

Gnome2 version 0.30
===================

The Gnome2 module allows a perl developer to use the Gnome libraries.
Find out more about Gnome+ at http://www.gnome.org.

The perl bindings follow the C API very closely, and the C reference
documentation should be considered the canonical source.

To discuss gtk2-perl, ask questions and flame/praise the authors,
join gtk-perl-list@gnome.org at lists.gnome.org.

Also have a look at the gtk2-perl website and sourceforge project page,
http://gtk2-perl.sourceforge.net

FIXME we have no other documentation, but we probably need it.


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install


To avoid installing to a system directory, since this is a beta release,
you can change the installation prefix at Makefile.PL time with

   perl Makefile.PL PREFIX=/some/other/place

This will install the module to the subdirectory lib/perl5 under the given
prefix.  If this is not already in perl's include path, you'll need to tell
perl how to get to this library directory so you can use it; there are three
ways:

  in your environment (the easiest):
    PERL5LIB=/some/other/place/lib/perl5/site_perl:$PERL5LIB

  on the perl command line:
    perl -I /some/other/place/lib/perl5/site_perl yourscript

  in the code of your perl script:
    use lib '/some/other/place/lib/perl5/site_perl';
  

DEPENDENCIES

This module requires these other modules and libraries:

  perl >= 5.8.0
  Glib >= 0.90 (perl module and requisite C libraries)
  Gtk2 >= 0.90 (perl module and requisite C libraries)
  libgnomeui > 2.x (C library)


COPYRIGHT AND LICENCE

Copyright (C) 2003 by the gtk2-perl team (see the file AUTHORS for a
complete list)

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; if not, write to the 
Free Software Foundation, Inc., 59 Temple Place - Suite 330, 
Boston, MA  02111-1307  USA.



NEWS
Overview of changes in Gnome2 1.046
===================================

* RT#104191; Don't use localtime() in POD output; patch submitted by intrigeri
  for the Debian Reproducable Builds project

Overview of changes in Gnome2 1.045
===================================

* Bugzilla #710417; Fix typo: correspoding -> corresponding

Overview of changes in Gnome2 1.044
===================================

* RT#89191: Fix FSF's address
* RT#89188: Make enums.pod depend on Gnome2 directory creation

Overview of changes in Gnome2 1.043
===================================

* Avoid misusing the macro PL_na, thus preventing issues when Gnome2 is used in
  conjunction with certain XS modules, among them XML::Parser and
  String::Approx.
* Created %meta_merge, which is used to pass META_MERGE values to EU::MM and
  CPAN::Meta; these populate META.[yml|json] and MYMETA.[yml|json]

Overview of changes in Gnome2 1.042
===================================

* Support the sm-connect property in Gnome2::Program::init.

Overview of changes in Gnome2 1.041
===================================

* Fix a test failure.
* Fix a POD error.

Overview of changes in Gnome2 1.040
===================================

 Since 1.02x (the previous stable series)
 ----------------------------------------
  * Fix some test suite issues.

 Since 1.030
 -----------
  * Nada.

Overview of changes in Gnome2 1.030
===================================

* Fix some test issues.

Overview of changes in Gnome2 1.023
===================================

* Fix yet another test suite issue.

Overview of changes in Gnome2 1.022
===================================

* Fix some test suite issues.

Overview of changes in Gnome2 1.021
===================================

* Update some build infrastructure.
* "Fix" a test failure.

Overview of changes in Gnome2 1.020
===================================

 Since 1.001 (the previous stable series)
 ----------------------------------------
  * Many internal changes and fixes to the test suite.
  * Possibility to use construct-only properties with
    Gnome2::Program::init.
  * A Gnome2::enums page that lists all enums and their values.
  * API additions:
    - Some new things in GnomePasswordDialog.
    - Gnome2::AuthenticationManager::dialog_is_visible.
    - Gnome2::IconList::select_all.

 Since 1.014
 -----------
  * Nada.

Overview of changes in Gnome2 1.014
===================================

* Bind and test gnome_authentication_manager_dialog_is_visible.
* Generate a Gnome2::enums man page with all enum and flag values.
  [Kevin C. Krinke]

Overview of changes in Gnome2 1.013
===================================

* Make construct-only properties work when using Gnome2::Program::init.

Overview of changes in Gnome2 1.012
===================================

* Fix man small glitches in the test suite; add a new test for GnomeSound.
* Bind and test new API in GnomePasswordDialog.

Overview of changes in Gnome2 1.011
===================================

* Bind and test Gnome2::IconList::select_all.
* Test Gnome2::DruidPageStandard::set_contents_background.

Overview of changes in Gnome2 1.010
===================================

* Rearrange many private functions and make them static.

Overview of changes in Gnome2 1.001
===================================

* Fix some building and testing issues.
* Fix a bug in Gnome2::IconTheme.
* Make Gnome2.pm use() Gnome2::Canvas so that Gnome2::IconList works.

Overview of changes in Gnome2 1.00
==================================

* Depend on Glib 1.04, Gnome2::Canvas 1.00, and Gnome2::VFS 1.00.

Overview of changes in Gnome2 0.94
==================================

* Document which functions may croak on error.

Overview of changes in Gnome2 0.93
==================================

* Changed the binding of gnome_popup_menu_append from Gtk2::Menu::append to
  Gtk2::Menu::append_from.  This is API breakage.  Any existing application
  using Gtk2::Menu::append and expecting to get gnome_popup_menu_append needs
  to be changed.
* Added new GnomePasswordDialog methods.
* Switched to the new version information policy introduced by Glib and Gtk2.
* Made minor bug fixes and improvements.
* In accordance with the GNOME Bindings release plan, Gnome2 froze API.


Gnome2.pm
# $Id$

package Gnome2;

use 5.008;
use strict;
use warnings;

use Gtk2;
use Gnome2::Canvas;
use Gnome2::VFS;

require DynaLoader;

our @ISA = qw(DynaLoader);

our $VERSION = '1.045';

sub import {
  my $self = shift();
  $self -> VERSION(@_);
}

sub dl_load_flags { $^O eq 'darwin' ? 0x00 : 0x01 }

Gnome2 -> bootstrap($VERSION);

1;
__END__

=head1 NAME

Gnome2 - Perl interface to the 2.x series of the GNOME libraries

=head1 SYNOPSIS

  use Gnome2;
  Gnome2::Program->init ($appname, $appversion);
  my $app = Gnome2::App->new ($appname);
  $app->show;
  Gtk2->main;

=head1 ABSTRACT

Perl bindings to the 2.x series of the GNOME libraries.  This module allows you
to interface with the libgnome and libgnomeui libraries to create GNOME
applications.

=head1 DESCRIPTION

Since this module tries to stick very closely to the C API, the documentation
found at

  http://developer.gnome.org/libgnome

and

  http://developer.gnome.org/libgnomeui

is the canonical reference.

In addition to that, there's also the automatically generated API
documentation: L<Gnome2::index>(3pm).

The mapping described in L<Gtk2::api>(3pm) also applies to this module.

To discuss this module, ask questions and flame/praise the authors, join
gtk-perl-list@gnome.org at lists.gnome.org.

=head1 MISSING METHODS

=over

=item gnome_app_fill_menu_with_data

=item gnome_app_fill_menu_custom

Use gnome_app_fill_menu instead.

=item gnome_app_fill_toolbar_with_data

=item gnome_app_fill_toolbar_custom

Use gnome_app_fill_toolbar instead.

=item gnome_app_create_menus_interp

=item gnome_app_create_menus_with_data

=item gnome_app_create_menus_custom

Use gnome_app_create_menus instead.

=item gnome_app_create_toolbar_interp

=item gnome_app_create_toolbar_with_data

=item gnome_app_create_toolbar_custom

Use gnome_app_create_toolbar instead.

=item gnome_app_insert_menus_custom

=item gnome_app_insert_menus_with_data

=item gnome_app_insert_menus_interp

Use gnome_app_insert_menus instead.

=back

=head1 RENAMED METHODS

=over

=item gnome_program_get

Gnome2::Program->get has been renamed to ->get_program to avoid a clash with
GObject->get.

=item gnome_popup_menu_attach

Gtk2::Menu->attach has been renamed to ->attach_to to avoid a clash with
gtk_menu_attach.

=item gnome_popup_menu_append

Gtk2::Menu->append has been renamed to ->append_from to avoid a clash with
gtk_menu_shell_append.

=back

=head1 DIFFERENT CALL SIGNATURES OR SEMANTICS

=over

=item gnome_client_set_restart_command

=item gnome_client_set_discard_command

=item gnome_client_set_resign_command

=item gnome_client_set_shutdown_command

=item gnome_client_set_clone_command

Instead of argc and argv you simply pass in a list.

=item gnome_program_init

Gnome2::Program->init automagically fills argc and argv so just omit these
parameters.

=back

=head1 SEE ALSO

L<perl>(1), L<Gnome2::index>(3pm), L<Glib>(3pm), L<Gtk2>(3pm),
L<Gtk2::api>(3pm).

=head1 AUTHOR

muppet E<lt>scott at asofyet dot orgE<gt> did the initial stuff.

Torsten Schoenfeld E<lt>kaffeetisch at gmx dot deE<gt> took over between 0.34
and 0.36, and implemented lots of tests and missing functionality, and
corrected several of muppet's mistakes.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2003-2006, 2013 by the gtk2-perl team (see the file AUTHORS)

This library is free software; you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published by the Free
Software Foundation; either version 2.1 of the License, or (at your option) any
later version.

This library is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

See the LICENSE file in the top level of this distribution for the complete
license terms.

=cut


gnome2perl.h
/*
 * Copyright (C) 2003 by the gtk2-perl team (see the file AUTHORS)
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * $Id$
 */

#ifndef _GNOME2PERL_H_
#define _GNOME2PERL_H_

#include <gtk2perl.h>
#include <vfs2perl.h>

#undef _ /* gnome and perl disagree on this one */

#include <gnome.h>
#include <libgnome/libgnometypebuiltins.h>

#include <libbonoboui.h>

#include "gnome2perl-versions.h"
#include "gnome2perl-autogen.h"

GnomeUIInfo *SvGnomeUIInfo (SV *sv);

void gnome2perl_parse_uiinfo_sv (SV * sv, GnomeUIInfo * info);
GnomeUIInfo * gnome2perl_svrv_to_uiinfo_tree (SV* sv, char * name);
void gnome2perl_refill_infos (SV *data, GnomeUIInfo *infos);
void gnome2perl_refill_infos_popup (SV *data, GnomeUIInfo *info);

#endif /* _GNOME2PERL_H_ */


Makefile.PL
#
# $Id$
#

use 5.008;
use strict;
use warnings;
use ExtUtils::MakeMaker;
use Cwd;
use File::Spec;

# minimum required version of dependencies we need to build
our %build_reqs = (
	'perl-ExtUtils-Depends'   => '0.20',
	'perl-ExtUtils-PkgConfig' => '1.03',
	'perl-Glib'               => '1.04',
	'perl-Gtk2'               => '1.00',
	'perl-Gnome2-Canvas'      => '1.00',
	'perl-Gnome2-VFS'         => '1.00',
	'Gnome'                   => '2.0.0',
	'Bonobo'                  => '2.0.0',
);

our %PREREQ_PM = (
	'Glib'                => $build_reqs{'perl-Glib'},
	'Gtk2'                => $build_reqs{'perl-Gtk2'},
	'Gnome2::Canvas'      => $build_reqs{'perl-Gnome2-Canvas'},
	'Gnome2::VFS'         => $build_reqs{'perl-Gnome2-VFS'},
	'ExtUtils::Depends'   => $build_reqs{'perl-ExtUtils-Depends'},
	'ExtUtils::PkgConfig' => $build_reqs{'perl-ExtUtils-PkgConfig'},
);

my %meta_merge = (
        q(meta-spec)          => {
            version => '2',
            url     => 'http://search.cpan.org/perldoc?CPAN::Meta::Spec',
        },
        author              =>
            ['Gnome2 Perl Team <gtk-perl-list at gnome dot org>'],
        release_status      => 'stable',
        # valid values: https://metacpan.org/module/CPAN::Meta::Spec#license
        license             => 'lgpl_2_1',
        resources => {
            license     => 'http://www.gnu.org/licenses/lgpl-2.1.html',
            homepage    => 'http://gtk2-perl.sourceforge.net',
            x_MailingList =>
                'https://mail.gnome.org/mailman/listinfo/gtk-perl-list',
            bugtracker  => {
                    web     =>
                    'http://rt.cpan.org/Public/Dist/Display.html?Name=Gnome2',
                    mailto  => 'bug-Gnome2 [at] rt.cpan.org',
            },
            repository  => {
                    url     => 'git://git.gnome.org/perl-Gnome2',
                    type    => 'git',
                    web     => 'http://git.gnome.org/browse/perl-Gnome2',
            },
        },
        prereqs => {
          configure => {
            requires => {%PREREQ_PM}, # no direct ref for 5.14 compatibility
          },
          test => {
            recommends => {
              'Test::Number::Delta' => 1.0,
            },
          },
        },
        no_index            => {
            file => [qw(inc/MakeHelper.pm)],
        },
);


# Writing a fake Makefile ensures that CPAN will pick up the correct
# dependencies and install them.
unless (eval "use ExtUtils::Depends '$build_reqs{'perl-ExtUtils-Depends'}';"
           . "use ExtUtils::PkgConfig '$build_reqs{'perl-ExtUtils-PkgConfig'}';"
           . "use Glib '$build_reqs{'perl-Glib'}';"
           . "use Gtk2 '$build_reqs{'perl-Gtk2'}';"
           . "use Gnome2::Canvas '$build_reqs{'perl-Gnome2-Canvas'}';"
           . "use Gnome2::VFS '$build_reqs{'perl-Gnome2-VFS'}';"
	   . "use Gtk2::CodeGen;"
           . "use Glib::MakeHelper;"
           . "1") {
   warn "$@\n";
   WriteMakefile(
         PREREQ_FATAL => 1,
         PREREQ_PM    => \%PREREQ_PM,
   );
   exit 1; # not reached
}

# now let's get started on the real work...
# in theory, getting what we need to build libgnomeui should 
# also get everything we need for libgnome.
my %libgnomeui_pkgcfg = ExtUtils::PkgConfig->find ("libgnomeui-2.0 >= $build_reqs{Gnome}");
my %libbonoboui_pkgcfg = ExtUtils::PkgConfig->find("libbonoboui-2.0 >= $build_reqs{Bonobo}");

mkdir 'build', 0777;

our @xs_files = <xs/*.xs>;
our %pod_files = (
	'Gnome2.pm' => '$(INST_MAN3DIR)/Gnome2.$(MAN3EXT)',
	# enums.pod is generated by rules we add to the postamble, below.
	# it is generated straight into blib to avoid dependency problems.
	'$(INST_LIB)/$(FULLEXT)/enums.pod'
	                   => '$(INST_MAN3DIR)/Gnome2::enums.$(MAN3EXT)',
	# we will be installing a whole slew of pod files, generated directly
	# from the xs source -- but to have them installed and manified
	# properly, we have to know their names at Makefile.PL time.
	#   This is a bit of a chicken and egg problem, and this solution
	# basically means that if you create a new package name in the xs
	# code, you'll need to re-run Makefile.PL.  you already need to
	# re-run to pick up new XS files, so that's not such a big deal. 
	Glib::MakeHelper->do_pod_files (@xs_files),
);

#
# autogeneration
#
Gtk2::CodeGen->parse_maps ('gnome2perl');
Gtk2::CodeGen->write_boot;

#
# create version guards for the libs that don't provide them.
#
ExtUtils::PkgConfig->write_version_macros(
	"build/gnome2perl-versions.h",
	'libgnome-2.0' => 'LIBGNOME',
	'libgnomeui-2.0' => 'LIBGNOMEUI',
	'libbonoboui-2.0' => 'LIBBONOBOUI',
);


# now we're ready to start creating the makefile.
# we need to use ExtUtils::Depends to get relevant information out of
# the Glib extension, and to save config information for other modules which
# will chain from this one.

our $gnome2 = ExtUtils::Depends->new ('Gnome2', 'Gnome2::Canvas', 'Gnome2::VFS');
$gnome2->set_inc ($libgnomeui_pkgcfg{cflags} . " " . $libbonoboui_pkgcfg{cflags});
$gnome2->set_libs ($libgnomeui_pkgcfg{libs} . " " . $libbonoboui_pkgcfg{libs});
$gnome2->add_xs (@xs_files);
$gnome2->add_pm ('Gnome2.pm' => '$(INST_LIBDIR)/Gnome2.pm');
$gnome2->add_typemaps (map {File::Spec->catfile(cwd(), $_)} 'build/gnome2perl.typemap', 'gnome.typemap');

$gnome2->install (qw(gnome2perl.h
                     build/gnome2perl-autogen.h
                     build/gnome2perl-versions.h
                     doctypes));
$gnome2->save_config ('build/IFiles.pm');

WriteMakefile(
    NAME          => 'Gnome2',
    VERSION_FROM  => 'Gnome2.pm', # finds $VERSION
    ABSTRACT_FROM => 'Gnome2.pm', # retrieve abstract from module
    XSPROTOARG    => '-noprototypes',
    MAN3PODS      => \%pod_files,
    PREREQ_PM     => \%PREREQ_PM,
    META_MERGE    => \%meta_merge,
    $gnome2->get_makefile_vars,
);

sub MY::postamble {
	return Glib::MakeHelper->postamble_clean ()
	     . Glib::MakeHelper->postamble_docs_full (
			DEPENDS => $main::gnome2,
			DOCTYPES => 'doctypes',
			COPYRIGHT_FROM => 'copyright.pod'
	       )
	     . Glib::MakeHelper->postamble_rpms (
			'GNOME' => $build_reqs{'Gnome'},
			'PERL_EXTUTILS_DEPENDS' =>
				$build_reqs{'perl-ExtUtils-Depends'},
			'PERL_EXTUTILS_PKGCONFIG' =>
				$build_reqs{'perl-ExtUtils-PkgConfig'},
			'PERL_GLIB' => $build_reqs{'perl-Glib'},
			'PERL_GTK' => $build_reqs{'perl-Gtk2'},
			'PERL_GNOME_CANVAS' => $build_reqs{'perl-Gnome2-Canvas'},
			'PERL_GNOME_VFS' => $build_reqs{'perl-Gnome2-VFS'},
		)
          . "
\$(INST_LIB)/\$(FULLEXT)/enums.pod : \$(BLIB_DONE) maps podifyenums.pl
	mkdir blib/lib/\$(NAME)
	\$(PERLRUNINST) -M\$(NAME) podifyenums.pl \$(NAME) maps > \$@
";
}

